src/sinks/util/encoding/encoding_config_with_default.rs

```diff
@@ -0,0 +1,197 @@
+use crate::sinks::util::encoding::{
+    EncodingConfig, EncodingConfiguration, InnerWithDefault, TimestampFormat,
+};
+use serde::de::{DeserializeOwned, IntoDeserializer, MapAccess, Visitor};
+use serde::{de, Deserialize, Deserializer, Serialize};
+use std::fmt;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use string_cache::DefaultAtom as Atom;
+
+/// A structure to wrap sink encodings and enforce field privacy.
+///
+/// This structure **does** assume that there is a default format. Consider
+/// `EncodingConfig<E>` instead if `E: !Default`.
+#[derive(Serialize, Debug, Eq, PartialEq, Clone, Default)]
+#[allow(dead_code)] // Required for `make check-component-features
+pub struct EncodingConfigWithDefault<E: Default + PartialEq> {
+    /// The format of the encoding.
+    // TODO: This is currently sink specific.
+    #[serde(
+        default,
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    pub(super) codec: E,
+    /// Keep only the following fields of the message. (Items mutually exclusive with `except_fields`)
+    #[serde(
+        default,
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    pub(super) only_fields: Option<Vec<Atom>>,
+    /// Remove the following fields of the message. (Items mutually exclusive with `only_fields`)
+    #[serde(
+        default,
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    pub(super) except_fields: Option<Vec<Atom>>,
+    /// Format for outgoing timestamps.
+    #[serde(
+        default,
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    pub(super) timestamp_format: Option<TimestampFormat>,
+}
+
+impl<E: Default + PartialEq> EncodingConfiguration<E> for EncodingConfigWithDefault<E> {
+    fn codec(&self) -> &E {
+        &self.codec
+    }
+    fn only_fields(&self) -> &Option<Vec<Atom>> {
+        &self.only_fields
+    }
+    fn set_only_fields(&mut self, fields: Option<Vec<Atom>>) -> Option<Vec<Atom>> {
+        std::mem::replace(&mut self.only_fields, fields)
+    }
+    fn except_fields(&self) -> &Option<Vec<Atom>> {
+        &self.except_fields
+    }
+    fn set_except_fields(&mut self, fields: Option<Vec<Atom>>) -> Option<Vec<Atom>> {
+        std::mem::replace(&mut self.only_fields, fields)
+    }
+    fn timestamp_format(&self) -> &Option<TimestampFormat> {
+        &self.timestamp_format
+    }
+    fn set_timestamp_format(&mut self, format: Option<TimestampFormat>) -> Option<TimestampFormat> {
+        std::mem::replace(&mut self.timestamp_format, format)
+    }
+}
+
+impl<E> EncodingConfigWithDefault<E>
+where
+    E: Default + PartialEq,
+{
+    #[allow(dead_code)] // Required for `make check-component-features`
```

Comment:
I think this could be `#[cfg(feature = "sinks-humio_logs")]` since that's the only user I can see. That's bound to explode though if others need it.

Reply:
I don't want to spend more time waiting for ``make check-component-features` than I have to so lets leave it.